### PR TITLE
[REF] Use Standard function cacheClause to re-use contact acl cache t…

### PR DESCRIPTION
**Pull request by seamuslee001**
…able

Overview
----------------------------------------
This refactors the CRM_Contact_BAO_Query functions to utilise the cacheClause function which joins on the acl_contact_cache table rather than re-running the ACL where clause again

Before
----------------------------------------
Doesn't use acl cache contact table

After
----------------------------------------
Uses ACL cache contact table

ping @kcristiano @eileenmcnaughton @colemanw

### CODE REVIEW

File: tests/phpunit/CRM/Contact/SelectorTest.php [hunk 1]
```diff
@@ -176,7 +176,7 @@ public function testPrevNextCache() {
       'expected_query' => [
         0 => 'default',
         1 => 'default',
-        2 => "WHERE  ( civicrm_email.email LIKE '%mickey@mouseville.com%' )  AND ( 1 )",
+        2 => "WHERE  ( civicrm_email.email LIKE '%mickey@mouseville.com%' )  AND contact_a.is_deleted = 0",
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contact/SelectorTest.php:179]: This is_deleted suggests the contact in the test does not have permission to view deleted contacts?

Comment (seamuslee001) [reply tests/phpunit/CRM/Contact/SelectorTest.php]: Looks like the cacheClause function was not respecting the view deleted contacts permission

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: CRM/Contact/BAO/Contact/Permission.php [hunk 2]
```diff
@@ -307,17 +307,22 @@ public static function cacheClause($contactAlias = 'contact_a') {
     if (CRM_Core_Permission::check('view all contacts') ||
       CRM_Core_Permission::check('edit all contacts')
     ) {
-      if (is_array($contactAlias)) {
-        $wheres = [];
-        foreach ($contactAlias as $alias) {
+      if (!CRM_Core_Permission::check('access deleted contacts')) {
+        if (is_array($contactAlias)) {
```
Comment (colemanw) [inline CRM/Contact/BAO/Contact/Permission.php:311]: I'm not a big fan of code that does `if(is_array($foo)) {.. do complex stuff multiple times} else {.. do complex stuff once}` if the "complex stuff" is the same in both cases, then it's better to just cast `$foo` to an array.

File: CRM/Contact/BAO/Contact/Permission.php [hunk 3]
```diff
@@ -307,17 +307,22 @@ public static function cacheClause($contactAlias = 'contact_a') {
...
+          $wheres = [];
+          foreach ($contactAlias as $alias) {
+            // CRM-6181
+            $wheres[] = "$alias.is_deleted = 0";
+          }
+          return [NULL, '(' . implode(' AND ', $wheres) . ')'];
+        }
```
Comment (colemanw) [inline CRM/Contact/BAO/Contact/Permission.php:318]:
```suggestion
        $wheres = [];
        foreach ((array) $contactAlias as $alias) {
          // CRM-6181
          $wheres[] = "$alias.is_deleted = 0";
        }
        return [NULL, '(' . implode(' AND ', $wheres) . ')'];
```

**Comment by eileenmcnaughton** [comment]
I added has-test as the previous fails make it clear this is tested!

@seamuslee001 it's stale

**Comment by seamuslee001** [comment]
Rebased now @eileenmcnaughton

**Comment by kcristiano** [comment]
@seamuslee001 This makes sense to me.   I'll add the PR to a couple of sites and see if I notice any oddities.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by eileenmcnaughton** [comment]
@kcristiano @colemanw this has been merge-ready for a while. I would normally merge something after merge-ready but this seems to be pending more testing rather than giving others a chance to input

**Comment by kcristiano** [comment]
@eileenmcnaughton I agree I have it applied on our 5.28 RC sites my concern is we won't hot any issues until I can get it on production.

I do think it's good to go in master as that will get it further testing before release

**Comment by eileenmcnaughton** [comment]
Thanks @kcristiano

**Pull request merged**